File: dockerfiles/attestation-sidecar-py/app/internal/blockchain/formatters.py
# app/internal/blockchain/formatters.py

from typing import Callable, Any, Sequence

from prettytable import PrettyTable
from web3 import Web3


def build_attestation_chain_table(
    attestation_chain: Sequence[bytes],
    get_attestation_info: Callable[[str], Any],
    get_agent_info: Callable[[str], Any],
    secaas_address: str = "0xed9d02e382b34818e88b88a309c7fe71e65f419d",
    last_n: int = 10,
) -> PrettyTable:
    """
    Returns a PrettyTable with the last N attestations.
    - get_attestation_info(attestation_id_str) -> (prover_address, verifier_address, attestation_result, timestamp)
    - get_agent_info(address) -> (name, is_registered, completed_attestations)
    """
    if not attestation_chain:
        table = PrettyTable()
        table.field_names = ["info"]
        table.add_row(["Attestation chain is empty"])
        return table

    last_n = min(last_n, len(attestation_chain))
    recent_chain = attestation_chain[-last_n:]

    table = PrettyTable()
    table.field_names = ["#", "Attestation ID", "Prover", "Verifier", "Result", "Timestamp"]
    table.align = "l"

    for i, attestation_id_b in enumerate(recent_chain, start=1):
        attestation_id = attestation_id_b.rstrip(b"\x00").decode("utf-8")

        prover_address, verifier_address, attestation_result, timestamp = get_attestation_info(attestation_id)
        status = "✅ SUCCESS" if attestation_result == 2 else "❌ FAILURE"

        prover_name, _, _ = get_agent_info(prover_address)

        if Web3.to_checksum_address(verifier_address) == Web3.to_checksum_address(secaas_address):
            verifier_name = "SECaaS"
        else:
            verifier_name, _, _ = get_agent_info(verifier_address)

        prover_info = f"{prover_name} - {prover_address}"
        verifier_info = f"{verifier_name} - {verifier_address}"

        table.add_row([i, attestation_id, prover_info, verifier_info, status, timestamp])

    return table

File: dockerfiles/attestation-sidecar/app/internal/blockchain/formatters.go
package blockchain

import (
	"bytes"
	"fmt"
	"math/big"
	"strings"

	"github.com/ethereum/go-ethereum/common"
	"github.com/olekukonko/tablewriter"
)

// Function types to match the Client methods
type GetAttestationInfoFunc func(string) (common.Address, common.Address, uint8, *big.Int, error)
type GetAgentInfoFunc func(common.Address) (string, bool, *big.Int, error)

// BuildAttestationChainTable generates a formatted ASCII table string for the chain.
func BuildAttestationChainTable(
	attestationChain [][32]byte,
	getAttestationInfo GetAttestationInfoFunc,
	getAgentInfo GetAgentInfoFunc,
	secaasAddressStr string,
	lastN int,
) string {
	// 1. Handle Empty Chain
	if len(attestationChain) == 0 {
		// Return a simple string or a table with one row, matching Python behavior
		var buf bytes.Buffer
		table := tablewriter.NewWriter(&buf)
		table.SetHeader([]string{"info"})
		table.Append([]string{"Attestation chain is empty"})
		table.Render()
		return buf.String()
	}

	// 2. Slice the last N items
	start := 0
	if len(attestationChain) > lastN {
		start = len(attestationChain) - lastN
	}
	recentChain := attestationChain[start:]

	// 3. Setup Table
	var buf bytes.Buffer
	table := tablewriter.NewWriter(&buf)
	table.SetHeader([]string{"#", "Attestation ID", "Prover", "Verifier", "Result", "Timestamp"})
	table.SetAlignment(tablewriter.ALIGN_LEFT)
	table.SetBorder(true) // Match PrettyTable's default look

	// Parse SECaaS address for comparison
	secaasAddr := common.HexToAddress(secaasAddressStr)
	if secaasAddressStr == "" {
		// Default from Python code
		secaasAddr = common.HexToAddress("0xed9d02e382b34818e88b88a309c7fe71e65f419d")
	}

	// 4. Iterate and Build Rows
	for i, idBytes := range recentChain {
		// Decode bytes32 to string and strip null bytes
		attID := strings.TrimRight(string(idBytes[:]), "\x00")

		// Fetch Info
		proverAddr, verifierAddr, attestationResult, timestamp, err := getAttestationInfo(attID)
		
		// If fetch fails, we render an error row to avoid crashing the whole table
		if err != nil {
			table.Append([]string{
				fmt.Sprintf("%d", i+1),
				attID,
				"ERROR",
				"ERROR",
				"UNKNOWN",
				"UNKNOWN",
			})
			continue
		}

		// Status Logic
		status := "❌ FAILURE"
		if attestationResult == 2 {
			status = "✅ SUCCESS"
		}

		// Prover Info
		proverName, _, _, _ := getAgentInfo(proverAddr)
		proverInfo := fmt.Sprintf("%s - %s", proverName, proverAddr.Hex())

		// Verifier Info
		verifierName := ""
		if verifierAddr == secaasAddr {
			verifierName = "SECaaS"
		} else {
			vName, _, _, _ := getAgentInfo(verifierAddr)
			verifierName = vName
		}
		verifierInfo := fmt.Sprintf("%s - %s", verifierName, verifierAddr.Hex())

		// Add Row
		table.Append([]string{
			fmt.Sprintf("%d", i+1),
			attID,
			proverInfo,
			verifierInfo,
			status,
			timestamp.String(),
		})
	}

	table.Render()
	return buf.String()
}